File: CourseBuddy.py
import json
import os

FEEDBACK_FILE = "feedback.json"

# Editable course catalog (full structure like your catalog variable)
COURSE_LIST = {
    "CPS100": {"name": "Computers and Applications", "course_number": "CPS100", "credits": 3, "prereqs": []},
    "CPS104": {"name": "Visual Programming", "course_number": "CPS104", "credits": 3, "prereqs": []},
    "CPS110": {"name": "Web Page Design", "course_number": "CPS110", "credits": 3, "prereqs": []},
    "CPS193": {"name": "Computer Science Selected Topic", "course_number": "CPS193", "credits": 1, "prereqs": []},
    "CPS210": {"name": "Computer Science I: Foundations", "course_number": "CPS210", "credits": 4, "prereqs": []},
    "CPS293": {"name": "Computer Science Selected Topic", "course_number": "CPS293", "credits": 1, "prereqs": []},
    "CPS295": {"name": "Independent Study Computer Science", "course_number": "CPS295", "credits": 1, "prereqs": []},
    "CPS296": {"name": "Departmental Elective", "course_number": "CPS296", "credits": 0, "prereqs": []},
    "CPS299": {"name": "Modular Course", "course_number": "CPS299", "credits": 0, "prereqs": []},
    "CPS310": {"name": "Computer Science II: Data Structures", "course_number": "CPS310", "credits": 4, "prereqs": ["CPS210"]},
    "CPS315": {"name": "Computer Science III", "course_number": "CPS315", "credits": 4, "prereqs": ["CPS310"]},
    "CPS330": {"name": "Assembly Language and Computer Architecture", "course_number": "CPS330", "credits": 4, "prereqs": ["CPS310"]},
    "CPS340": {"name": "Operating Systems", "course_number": "CPS340", "credits": 4, "prereqs": ["CPS330"]},
    "CPS341": {"name": "Operating Systems II", "course_number": "CPS341", "credits": 3, "prereqs": ["CPS340"]},
    "CPS342": {"name": "Embedded Linux", "course_number": "CPS342", "credits": 3, "prereqs": ["CPS310"]},
    "CPS352": {"name": "Object Oriented Programming", "course_number": "CPS352", "credits": 3, "prereqs": ["CPS310"]},
    "CPS353": {"name": "Software Engineering", "course_number": "CPS353", "credits": 3, "prereqs": ["CPS310"]},
    "CPS393": {"name": "Computer Science Selected Topic", "course_number": "CPS393", "credits": 1, "prereqs": []},
    "CPS396": {"name": "Departmental Elective", "course_number": "CPS396", "credits": 0, "prereqs": []},
    "CPS399": {"name": "Modular Course", "course_number": "CPS399", "credits": 0, "prereqs": []},
    "CPS415": {"name": "Discrete and Continuous Computer Algorithms", "course_number": "CPS415", "credits": 3, "prereqs": ["MAT320"]},
    "CPS425": {"name": "Language Processing", "course_number": "CPS425", "credits": 4, "prereqs": ["CPS310","CPS330"]},
    "CPS440": {"name": "Database Principles", "course_number": "CPS440", "credits": 3, "prereqs": ["CPS310"]},
    "CPS441": {"name": "Database Projects", "course_number": "CPS441", "credits": 4, "prereqs": ["CPS440"]},
    "CPS460": {"name": "Computer Architecture", "course_number": "CPS460", "credits": 3, "prereqs": []},
    "CPS470": {"name": "Computer Communication Networks", "course_number": "CPS470", "credits": 3, "prereqs": []},
    "CPS471": {"name": "Computer Communication Networks II", "course_number": "CPS471", "credits": 4, "prereqs": ["CPS470"]},
    "CPS485": {"name": "Projects", "course_number": "CPS485", "credits": 4, "prereqs": ["CPS493","CPS470","CPS440"]},
    "CPS493": {"name": "Computer Science Selected Topic", "course_number": "CPS493", "credits": 3, "prereqs": []},
    "CPS494": {"name": "Fieldwork Computer Science", "course_number": "CPS494", "credits": 1, "prereqs": []},
    "CPS495": {"name": "Independent Study Computer Science", "course_number": "CPS495", "credits": 1, "prereqs": []},
    "MAT251": {"name": "Calculus I", "course_number": "MAT251", "credits": 4, "prereqs": ["MAT181"]},
    "MAT252": {"name": "Calculus II", "course_number": "MAT252", "credits": 4, "prereqs": ["MAT251"]},
    "MAT320": {"name": "Discrete Mathematics for Computing", "course_number": "MAT320", "credits": 3, "prereqs": ["MAT181"]},
    "EGC220": {"name": "Digital Logic Fundamentals", "course_number": "EGC220", "credits": 3, "prereqs": ["MAT251", "EGC221"]},
    "EGS221": {"name": "Digital Logic Lab", "course_number": "EGC221", "credits": 1, "prereqs": ["EGC220"]},
}




# ---------- Feedback System ----------

def load_feedback():
    """Load existing feedback from file."""
    if not os.path.exists(FEEDBACK_FILE):
        return {}
    with open(FEEDBACK_FILE, "r") as file:
        return json.load(file)


def save_feedback(data):
    """Save feedback to JSON file."""
    with open(FEEDBACK_FILE, "w") as file:
        json.dump(data, file, indent=4)


def give_advice(feedback_data):
    """Display advice for a selected course."""
    course = input("Enter the course number you want advice on (e.g., CPS352): ").upper()

    if course not in COURSE_LIST:
        print("❌ This course is not recognized. Please check your input.")
        return

    course_info = COURSE_LIST[course]
    print(f"\nCourse Selected: {course} - {course_info['name']}")
    print(f"Credits: {course_info['credits']}")
    print(f"Prerequisites: {', '.join(course_info['prereqs']) if course_info['prereqs'] else 'None'}")

    if course not in feedback_data:
        print("\nNo prior student feedback available yet. Proceed with standard preparation.")
        return

    print("\n--- Student Feedback Summary ---")
    for entry in feedback_data[course]:
        print(f"• Reported Assignment/Exam Issues: {entry['problems']}")
        print(f"• Recommended Prerequisite(s): {entry['recommended_prereq']}\n")


def collect_feedback(feedback_data):
    """Collect new feedback from the user."""
    course = input("Enter the course number you are giving feedback for (e.g., CPS352): ").upper()

    if course not in COURSE_LIST:
        print("❌ This course is not recognized. Please check your input.")
        return

    problems = input("What were the problems with exams and assignments? ")
    recommended_prereq = input("Recommended course(s) you should take before this one (list course numbers): ")

    if course not in feedback_data:
        feedback_data[course] = []

    feedback_data[course].append({
        "problems": problems,
        "recommended_prereq": recommended_prereq
    })

    save_feedback(feedback_data)
    print("\n✅ Your feedback has been saved. Thank you!\n")


def main():
    feedback_data = load_feedback()

    print("Welcome to the Computer Science Course Advisor & Feedback System\n")
    choice = input("Do you want advice (A) or give feedback (F)? ").strip().upper()

    if choice == "A":
        give_advice(feedback_data)
    elif choice == "F":
        collect_feedback(feedback_data)
    else:
        print("Invalid option. Please restart the program.")


if __name__ == "__main__":
    main()

File: CourseBuddy.java
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.File;
import java.io.FileWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class CourseBuddy {

    private static final String FEEDBACK_FILE = "feedback.json";

    // Editable course list
    private static final Map<String, String> COURSE_LIST = new HashMap<>();
    static {
        COURSE_LIST.put("CPS210", "Computer Science I");
        COURSE_LIST.put("CPS310", "Computer Science II");
        COURSE_LIST.put("CPS315", "Computer Organization");
        COURSE_LIST.put("CPS330", "Assembly Programming");
        COURSE_LIST.put("CPS352", "Data Structures");
        COURSE_LIST.put("CPS353", "Operating Systems");
        COURSE_LIST.put("CPS430", "Database Systems");
        COURSE_LIST.put("CPS493", "Computer Science Seminar");
    }

    // Load feedback from JSON file
    private static JSONObject loadFeedback() {
        try {
            File file = new File(FEEDBACK_FILE);
            if (!file.exists()) {
                return new JSONObject();
            }
            String content = new String(Files.readAllBytes(Paths.get(FEEDBACK_FILE)));
            return new JSONObject(content);
        } catch (Exception e) {
            e.printStackTrace();
            return new JSONObject();
        }
    }

    // Save JSON feedback to file
    private static void saveFeedback(JSONObject data) {
        try (FileWriter file = new FileWriter(FEEDBACK_FILE)) {
            file.write(data.toString(4));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    // Give advice based on previous feedback
    private static void giveAdvice(JSONObject feedbackData, Scanner scanner) {
        System.out.print("Enter the course number you want advice on (e.g., CPS352): ");
        String course = scanner.nextLine().trim().toUpperCase();

        if (!COURSE_LIST.containsKey(course)) {
            System.out.println("This course is not recognized. Make sure you entered it correctly.");
            return;
        }

        System.out.println("\nCourse Selected: " + course + " - " + COURSE_LIST.get(course));

        if (!feedbackData.has(course)) {
            System.out.println("No prior student feedback available yet. Proceed with standard preparation.");
            return;
        }

        System.out.println("\n--- Student Feedback Summary ---");
        JSONArray feedbackArray = feedbackData.getJSONArray(course);
        for (int i = 0; i < feedbackArray.length(); i++) {
            JSONObject entry = feedbackArray.getJSONObject(i);
            System.out.println("• Reported Assignment/Exam Issues: " + entry.getString("problems"));
            System.out.println("• Recommended Prerequisite(s): " + entry.getString("recommended_prereq") + "\n");
        }
    }

    // Collect new feedback
    private static void collectFeedback(JSONObject feedbackData, Scanner scanner) {
        System.out.print("Enter the course number you are giving feedback for (e.g., CPS352): ");
        String course = scanner.nextLine().trim().toUpperCase();

        System.out.print("What were the problems with exams and assignments? ");
        String problems = scanner.nextLine();

        System.out.print("Recommended course(s) you should take before this one: ");
        String recommendedPrereq = scanner.nextLine();

        JSONArray feedbackArray = feedbackData.optJSONArray(course);
        if (feedbackArray == null) {
            feedbackArray = new JSONArray();
            feedbackData.put(course, feedbackArray);
        }

        JSONObject entry = new JSONObject();
        entry.put("problems", problems);
        entry.put("recommended_prereq", recommendedPrereq);
        feedbackArray.put(entry);

        saveFeedback(feedbackData);
        System.out.println("\n✅ Your feedback has been saved. Thank you!\n");
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        JSONObject feedbackData = loadFeedback();

        System.out.println("Welcome to the Computer Science Course Advisor & Feedback System\n");
        System.out.print("Do you want advice (A) or give feedback (F)? ");
        String choice = scanner.nextLine().trim().toUpperCase();

        if (choice.equals("A")) {
            giveAdvice(feedbackData, scanner);
        } else if (choice.equals("F")) {
            collectFeedback(feedbackData, scanner);
        } else {
            System.out.println("Invalid option. Please restart the program.");
        }
    }
}
